src/main/java/com/mongodb/spark/sql/connector/read/partitioner/PartitionerHelper.java

```diff
@@ -0,0 +1,157 @@
+/*
+ * Copyright 2008-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package com.mongodb.spark.sql.connector.read.partitioner;
+
+import static com.mongodb.spark.sql.connector.read.partitioner.Partitioner.LOGGER;
+import static java.util.Collections.singletonList;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
+
+import org.bson.BsonDocument;
+import org.bson.BsonType;
+import org.bson.BsonValue;
+
+import com.mongodb.MongoCommandException;
+import com.mongodb.client.MongoClient;
+
+import com.mongodb.spark.sql.connector.config.ReadConfig;
+import com.mongodb.spark.sql.connector.exceptions.MongoSparkException;
+import com.mongodb.spark.sql.connector.read.MongoInputPartition;
+
+/** The Partitioner helper class */
+public final class PartitionerHelper {
+
+  private static final List<BsonDocument> COLL_STATS_AGGREGATION_PIPELINE =
+      singletonList(BsonDocument.parse("{'$collStats': {'storageStats': { } } }"));
+  public static final String DEFAULT_PARTITIONER = SamplePartitioner.class.getCanonicalName();
+  public static final Partitioner SINGLE_PARTITIONER = new SinglePartitionPartitioner();
+
+  /**
+   * @param readConfig the read config
+   * @return the partitioner class name
+   */
+  public static MongoInputPartition[] generatePartitions(final ReadConfig readConfig) {
+    try {
+      Partitioner partitioner = readConfig.getPartitioner();
+      LOGGER.debug("Generating partitions using '{}'.", partitioner.getClass().getSimpleName());
+      List<MongoInputPartition> mongoInputPartitions = partitioner.generatePartitions(readConfig);
+      LOGGER.debug(
+          "Partitioner '{}' created {} partition(s).",
+          partitioner.getClass().getSimpleName(),
+          mongoInputPartitions.size());
+
+      if (mongoInputPartitions.isEmpty()) {
+        LOGGER.warn(
+            "Partitioner '{}' failed to create any partitions. Falling back to a single partition for the collection",
+            partitioner.getClass().getSimpleName());
+        mongoInputPartitions = SINGLE_PARTITIONER.generatePartitions(readConfig);
+      }
+
+      return mongoInputPartitions.toArray(new MongoInputPartition[0]);
+    } catch (RuntimeException ex) {
+      LOGGER.error("Partitioning failed due to {}", ex.getMessage());
+      throw new MongoSparkException("Partitioning failed.", ex);
```

Comment:
Its because we can have user provided partitioners which may be buggy.  

The extra logging step is to provide assistance when debugging in support instances - rather than a chaotic stacktrace. Also wrapping the exception further provides assistance.

Reply:
> The extra logging step is to provide assistance when debugging in support instances - rather than a chaotic stacktrace

We supply no new useful information in the logging message. All we do is repeat the exception message (`ex.getMessage()`) that is duplicated when the exception is logged. This way we not only clutter the code, but we also clutter logs, which is worse.

Reply:
Ack